lib/srv/mcp/audit.go

```diff
@@ -0,0 +1,228 @@
+		sessionAuditorConfig: cfg,
+	}, nil
+}
+
+func (a *sessionAuditor) shouldEmitEvent(method mcp.MCPMethod) bool {
+	// Do not record discovery, ping calls.
+	switch method {
+	case mcp.MethodPing,
+		mcp.MethodResourcesList,
+		mcp.MethodResourcesTemplatesList,
+		mcp.MethodPromptsList,
+		mcp.MethodToolsList:
+		return false
+	default:
+		return true
+	}
+}
+
+func (a *sessionAuditor) emitStartEvent(ctx context.Context) {
+	a.emitEvent(ctx, &apievents.MCPSessionStart{
+		Metadata: a.makeEventMetadata(
+			events.MCPSessionStartEvent,
+			events.MCPSessionStartCode,
+		),
+		ServerMetadata:     a.makeServerMetadata(),
+		SessionMetadata:    a.makeSessionMetadata(),
+		UserMetadata:       a.makeUserMetadata(),
+		ConnectionMetadata: a.makeConnectionMetadata(),
+		AppMetadata:        a.makeAppMetadata(),
+	})
+}
+
+func (a *sessionAuditor) emitEndEvent(ctx context.Context) {
+	a.emitEvent(ctx, &apievents.MCPSessionEnd{
+		Metadata: a.makeEventMetadata(
+			events.MCPSessionEndEvent,
+			events.MCPSessionEndCode,
+		),
+		ServerMetadata:     a.makeServerMetadata(),
+		SessionMetadata:    a.makeSessionMetadata(),
+		UserMetadata:       a.makeUserMetadata(),
+		ConnectionMetadata: a.makeConnectionMetadata(),
+		AppMetadata:        a.makeAppMetadata(),
+	})
+}
+
+func (a *sessionAuditor) emitNotificationEvent(ctx context.Context, msg *mcputils.JSONRPCNotification) {
+	if !a.shouldEmitEvent(msg.Method) {
+		return
+	}
+	a.emitEvent(ctx, &apievents.MCPSessionNotification{
+		Metadata: a.makeEventMetadata(
+			events.MCPSessionNotificationEvent,
+			events.MCPSessionNotificationCode,
+		),
+		SessionMetadata: a.makeSessionMetadata(),
+		UserMetadata:    a.makeUserMetadata(),
+		AppMetadata:     a.makeAppMetadata(),
+		Message: apievents.MCPJSONRPCMessage{
+			JSONRPC: msg.JSONRPC,
+			Method:  string(msg.Method),
+			Params:  msg.Params.GetEventParams(),
+		},
+	})
+}
+
+func (a *sessionAuditor) emitRequestEvent(ctx context.Context, msg *mcputils.JSONRPCRequest, err error) {
+	if err == nil && !a.shouldEmitEvent(msg.Method) {
+		return
+	}
+	event := &apievents.MCPSessionRequest{
+		Metadata: a.makeEventMetadata(
+			events.MCPSessionRequestEvent,
+			events.MCPSessionRequestCode,
+		),
+		SessionMetadata: a.makeSessionMetadata(),
+		UserMetadata:    a.makeUserMetadata(),
+		AppMetadata:     a.makeAppMetadata(),
+		Status: apievents.Status{
+			Success: true,
+		},
+		Message: apievents.MCPJSONRPCMessage{
+			JSONRPC: msg.JSONRPC,
+			Method:  string(msg.Method),
+			ID:      msg.ID.String(),
+			Params:  msg.Params.GetEventParams(),
+		},
+	}
+
+	if err != nil {
+		event.Metadata.Code = events.MCPSessionRequestFailureCode
+		event.Status.Success = false
+		event.Status.Error = err.Error()
+	}
+	a.emitEvent(ctx, event)
+}
+
+func (a *sessionAuditor) emitEvent(ctx context.Context, event apievents.AuditEvent) {
+	preparedEvent, err := a.preparer.PrepareSessionEvent(event)
+	if err != nil {
+		a.logger.ErrorContext(ctx, "Failed to setup event",
```

Comment:
nit: for consistency with naming, unless this setup is more correct for some other reason
```suggestion
		a.logger.ErrorContext(ctx, "Failed to prepare event",
```